registry/storage/validatorstore.go

```diff
@@ -0,0 +1,366 @@
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	// Update byValidatorIndex
+	for _, share := range shares {
+		if share.HasBeaconMetadata() {
+			c.byValidatorIndex[share.BeaconMetadata.Index] = share
+		}
+
+		// Update byCommitteeID
+		committee := c.byCommitteeID[share.CommitteeID()]
+		if committee == nil {
+			committee = buildCommittee([]*types.SSVShare{share})
+			c.byCommitteeID[committee.ID] = committee
+		} else {
+			committee = buildCommittee(append(committee.Validators, share))
+		}
+		c.byCommitteeID[committee.ID] = committee
+
+		// Update byOperatorID
+		for _, operator := range share.Committee {
+			data := c.byOperatorID[operator.Signer]
+			if data == nil {
+				data = &sharesAndCommittees{
+					shares:     []*types.SSVShare{share},
+					committees: []*Committee{committee},
+				}
+			} else {
+				data.shares = append(data.shares, share)
+				data.committees = append(data.committees, committee)
+			}
+			c.byOperatorID[operator.Signer] = data
+		}
+	}
+}
+
+func (c *validatorStore) handleShareRemoved(pk spectypes.ValidatorPK) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	share := c.byPubKey(pk[:])
+	if share == nil {
+		return
+	}
+
+	// Update byValidatorIndex
+	delete(c.byValidatorIndex, share.BeaconMetadata.Index)
+
+	// Update byCommitteeID
+	committee := c.byCommitteeID[share.CommitteeID()]
+	if committee == nil {
+		return
+	}
+	validators := make([]*types.SSVShare, 0, len(committee.Validators)-1)
+	indices := make([]phase0.ValidatorIndex, 0, len(committee.Validators)-1)
+	for _, validator := range committee.Validators {
+		if validator.ValidatorPubKey != pk {
+			validators = append(validators, validator)
+			indices = append(indices, validator.ValidatorIndex)
+		}
+	}
+	if len(validators) == 0 {
+		delete(c.byCommitteeID, committee.ID)
+	} else {
+		committee.Validators = validators
+		committee.Indices = indices
+	}
+
+	// Update byOperatorID
+	for _, operator := range share.Committee {
+		data := c.byOperatorID[operator.Signer]
+		if data == nil {
+			return
+		}
+		shares := make([]*types.SSVShare, 0, len(data.shares)-1)
+		for _, s := range data.shares {
+			if s.ValidatorPubKey != pk {
+				shares = append(shares, s)
+			}
+		}
+		if len(shares) == 0 {
+			delete(c.byOperatorID, operator.Signer)
+		} else {
+			data.shares = shares
+		}
+	}
+}
+
+func (c *validatorStore) handleShareUpdated(share *types.SSVShare) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	// Update byValidatorIndex
+	if share.HasBeaconMetadata() {
+		c.byValidatorIndex[share.BeaconMetadata.Index] = share
+	}
+
+	// Update byCommitteeID
+	for _, committee := range c.byCommitteeID {
+		for i, validator := range committee.Validators {
+			if validator.ValidatorPubKey == share.ValidatorPubKey {
+				committee.Validators[i] = share
+				break
+			}
+		}
+
+		for i, index := range committee.Indices {
+			if index == share.ValidatorIndex {
+				committee.Indices[i] = share.ValidatorIndex
+				break
+			}
+		}
```

Comment:
@nkryuchkov this doesn't seem right. how can this work? 
if the index is already the correct index from the share why we even need to run this code?
if the index isn't correct - on fresh sync before the metadata update, the index will be 0. then this update code just won't work.

Probably best to set this above, where you found the validator pubkey, and use the same slice index to update the validator index.

Reply:
https://github.com/ssvlabs/ssv/pull/1495